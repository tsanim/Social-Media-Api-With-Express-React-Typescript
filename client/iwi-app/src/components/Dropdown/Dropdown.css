.dropbtn {
    color: white;
    padding: 16px;
    font-size: 16px;
}

.dropdown {
    position: relative;
    display: inline-block;
}

.dropdown-content {
    display: none;
    position: absolute;
    background-color: #8AAAE5;
    min-width: 160px;
    box-shadow: 0px 8px 16px 0px rgba(0, 0, 0, 0.2);
    z-index: 1;
}

.dropdown-content a {
    color: black;
    padding: 12px 16px;
    text-decoration: none;
    display: block;
}

.dropdown:hover .dropdown-content {
    display: block;
}

.dropdown:hover .dropbtn {
    color: rgb(65, 65, 124);
}