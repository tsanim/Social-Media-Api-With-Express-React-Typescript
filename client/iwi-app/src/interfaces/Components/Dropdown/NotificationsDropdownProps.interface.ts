import Notification from "../../Feed/Notification.interface";

export default interface NotificationsDropdownProps {
    notifications: Notification[];
}