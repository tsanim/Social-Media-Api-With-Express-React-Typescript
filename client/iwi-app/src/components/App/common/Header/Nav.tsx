import React from 'react';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faHome, faUsers, faSignOutAlt, faImages } from '@fortawesome/free-solid-svg-icons';
import { Detector } from "react-detect-offline";
import PropTypes from 'prop-types';

import URI from '../../../../config/config';
import Dropdown from '../../../Dropdown/NotificationsDropdown';
import HeaderProps from '../../../../interfaces/Components/Header/HeaderProps.interface';

function Nav(props: HeaderProps) {
    const { imageId } = props.user;

    return (
        <nav id="main-nav">
            <ul>
                <Detector
                    // tslint:disable-next-line: jsx-no-multiline-js
                    render={({ online }: { online: boolean}) => {
                        if (online) {
                            props.switchToOnline();
                        } else {
                            props.switchToOffline();
                        }

                        return (
                            
                            <li className={`status ${online ? 'online' : 'offline'}`}/>
                        )
                    }}
                />
                <li>
                    <Link to="/MyProfile">
                        <figure className="nav-profilePic">
                            <img src={`${URI}/feed/image/${imageId}`} alt="" />
                            <figcaption>{localStorage.getItem('username')}</figcaption>
                        </figure>
                    </Link>
                </li>
                <li><Link to="/"><FontAwesomeIcon icon={faHome} /> HOME</Link></li>
                <li><Link to="/discover"><FontAwesomeIcon icon={faUsers} /> DISCOVER</Link></li>
                <li><Link to="/searchPosts"><FontAwesomeIcon icon={faImages} /> SEARCH</Link></li>
                <li> <Dropdown
                    notifications={props.notifications}
                />
                </li>
                <li><Link to="/chat"><FontAwesomeIcon icon={faUsers} /> CHAT</Link></li>
                <li><Link to="/signin" onClick={props.signoutHandler}><FontAwesomeIcon icon={faSignOutAlt} /> LOGOUT</Link></li>
            </ul>
        </nav>
    )
}

Nav.propTypes = {
    user: PropTypes.object,
    resetUserPostsHandler: PropTypes.func,
    signoutHandler: PropTypes.func,
    switchToOffline: PropTypes.func,
    switchToOnline: PropTypes.func,
}

export default Nav;