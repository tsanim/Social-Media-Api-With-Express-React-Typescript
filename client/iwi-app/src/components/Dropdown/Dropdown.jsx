import React from 'react';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import './Dropdown.css';
import { faBell } from '@fortawesome/free-solid-svg-icons';
import PropTypes from 'prop-types';

function Dropdown(props) {
    return (
        <div className="dropdown">
            <button className="dropbtn"><FontAwesomeIcon icon={faBell} />{props.notifications.length}</button>
            <div className="dropdown-content">
                {
                    props.notifications.map(n => <Link key={n._id} to={`/chat?sender=${n.sender}`}> {n.message} </Link>)
                }
            </div>
        </div>
    )
}

Dropdown.propTypes = {
    notifications: PropTypes.array,
}

export default Dropdown;