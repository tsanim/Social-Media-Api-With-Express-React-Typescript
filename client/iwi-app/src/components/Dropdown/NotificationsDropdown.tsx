import React from 'react';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import './Dropdown.css';
import { faBell } from '@fortawesome/free-solid-svg-icons';
import PropTypes from 'prop-types';
import DropdownProps from '../../interfaces/Components/Dropdown/NotificationsDropdownProps.interface';

function Dropdown(props: DropdownProps) {
    return (
        <div className="dropdown">
            <button className="dropbtn"><FontAwesomeIcon icon={faBell} />{props.notifications.length}</button>
            <div className="dropdown-content">
                {
                    props.notifications.map(n => <Link key={n._id} to={`/chat?sender=${n.sender}`}> {n.message} </Link>)
                }
            </div>
        </div>
    )
}

Dropdown.propTypes = {
    notifications: PropTypes.array,
}

export default Dropdown;