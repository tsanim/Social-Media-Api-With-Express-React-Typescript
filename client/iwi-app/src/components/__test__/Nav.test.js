import React from 'react';
import Nav from '../App/common/Header/Nav';
import { Detector } from "react-detect-offline";
import Dropdown from '../Dropdown/Dropdown';

import { shallow } from 'enzyme';

function navComponentSetup() {
    const props = {
        user: {},
        resetUserPostsHandler: jest.fn(),
        signoutHandler: jest.fn(),
        switchToOffline: jest.fn(),
        switchToOnline: jest.fn()
    }

    window.localStorage.setItem('username', 'Tsani');

    const enzymeWrapper = shallow(<Nav {...props} />)

    return {
        props,
        enzymeWrapper
    }
}

describe('Header component', () => {
    it('should render initial layout', () => {
        const { enzymeWrapper } = navComponentSetup();

        expect(enzymeWrapper.getElements()).toMatchSnapshot();
    });

    it('should render nav tag container with id "main-nav"', () => {
        const { enzymeWrapper } = navComponentSetup();

        expect(enzymeWrapper.find('nav#main-nav')).toHaveLength(1);
    });

    it('should render ul tag', () => {
        const { enzymeWrapper } = navComponentSetup();

        expect(enzymeWrapper.find('ul')).toHaveLength(1);
    });

    it('should render Detector Component', () => {
        const { enzymeWrapper } = navComponentSetup();

        expect(enzymeWrapper.find(Detector)).toHaveLength(1);
    });

    it('should render ul tag with 8 items', () => {
        const { enzymeWrapper } = navComponentSetup();

        expect(enzymeWrapper.find('ul').children()).toHaveLength(8);
    });

    it('should render Dropdown Component', () => {
        const { enzymeWrapper } = navComponentSetup();

        expect(enzymeWrapper.find(Dropdown)).toHaveLength(1);
    });

    it('should render Detector Component with li tag with class "status online" if status online is true', () => {
        const { enzymeWrapper } = navComponentSetup();
        const hasClass = enzymeWrapper.find(Detector).renderProp('render')({ online: true }).find('li').hasClass('status online');

        expect(hasClass).toEqual(true);
    });

    it('should render Detector Component with li tag with class "status offline" if status online is false', () => {
        const { enzymeWrapper } = navComponentSetup();
        const hasClass = enzymeWrapper.find(Detector).renderProp('render')({ online: false }).find('li').hasClass('status offline');

        expect(hasClass).toEqual(true);
    });
});