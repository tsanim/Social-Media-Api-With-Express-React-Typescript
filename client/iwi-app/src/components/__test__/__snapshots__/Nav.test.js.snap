// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render initial layout 1`] = `
Array [
  <nav
    id="main-nav"
  >
    <ul>
      <Detector
        polling={true}
        render={[Function]}
        wrapperType="span"
      />
      <li>
        <Link
          to="/MyProfile"
        >
          <figure
            className="nav-profilePic"
          >
            <img
              alt=""
              src="http://localhost:9999/feed/image/undefined"
            />
            <figcaption>
              Tsani
            </figcaption>
          </figure>
        </Link>
      </li>
      <li>
        <Link
          to="/"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  576,
                  512,
                  Array [],
                  "f015",
                  "M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z",
                ],
                "iconName": "home",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
           HOME
        </Link>
      </li>
      <li>
        <Link
          to="/discover"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  640,
                  512,
                  Array [],
                  "f0c0",
                  "M96 224c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm448 0c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm32 32h-64c-17.6 0-33.5 7.1-45.1 18.6 40.3 22.1 68.9 62 75.1 109.4h66c17.7 0 32-14.3 32-32v-32c0-35.3-28.7-64-64-64zm-256 0c61.9 0 112-50.1 112-112S381.9 32 320 32 208 82.1 208 144s50.1 112 112 112zm76.8 32h-8.3c-20.8 10-43.9 16-68.5 16s-47.6-6-68.5-16h-8.3C179.6 288 128 339.6 128 403.2V432c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48v-28.8c0-63.6-51.6-115.2-115.2-115.2zm-223.7-13.4C161.5 263.1 145.6 256 128 256H64c-35.3 0-64 28.7-64 64v32c0 17.7 14.3 32 32 32h65.9c6.3-47.4 34.9-87.3 75.2-109.4z",
                ],
                "iconName": "users",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
           DISCOVER
        </Link>
      </li>
      <li>
        <Link
          to="/searchPosts"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  576,
                  512,
                  Array [],
                  "f302",
                  "M480 416v16c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V176c0-26.51 21.49-48 48-48h16v208c0 44.112 35.888 80 80 80h336zm96-80V80c0-26.51-21.49-48-48-48H144c-26.51 0-48 21.49-48 48v256c0 26.51 21.49 48 48 48h384c26.51 0 48-21.49 48-48zM256 128c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-96 144l55.515-55.515c4.686-4.686 12.284-4.686 16.971 0L272 256l135.515-135.515c4.686-4.686 12.284-4.686 16.971 0L512 208v112H160v-48z",
                ],
                "iconName": "images",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
           SEARCH
        </Link>
      </li>
      <li>
         
        <Dropdown />
      </li>
      <li>
        <Link
          to="/chat"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  640,
                  512,
                  Array [],
                  "f0c0",
                  "M96 224c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm448 0c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm32 32h-64c-17.6 0-33.5 7.1-45.1 18.6 40.3 22.1 68.9 62 75.1 109.4h66c17.7 0 32-14.3 32-32v-32c0-35.3-28.7-64-64-64zm-256 0c61.9 0 112-50.1 112-112S381.9 32 320 32 208 82.1 208 144s50.1 112 112 112zm76.8 32h-8.3c-20.8 10-43.9 16-68.5 16s-47.6-6-68.5-16h-8.3C179.6 288 128 339.6 128 403.2V432c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48v-28.8c0-63.6-51.6-115.2-115.2-115.2zm-223.7-13.4C161.5 263.1 145.6 256 128 256H64c-35.3 0-64 28.7-64 64v32c0 17.7 14.3 32 32 32h65.9c6.3-47.4 34.9-87.3 75.2-109.4z",
                ],
                "iconName": "users",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
           CHAT
        </Link>
      </li>
      <li>
        <Link
          onClick={[MockFunction]}
          to="/signin"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  512,
                  512,
                  Array [],
                  "f2f5",
                  "M497 273L329 441c-15 15-41 4.5-41-17v-96H152c-13.3 0-24-10.7-24-24v-96c0-13.3 10.7-24 24-24h136V88c0-21.4 25.9-32 41-17l168 168c9.3 9.4 9.3 24.6 0 34zM192 436v-40c0-6.6-5.4-12-12-12H96c-17.7 0-32-14.3-32-32V160c0-17.7 14.3-32 32-32h84c6.6 0 12-5.4 12-12V76c0-6.6-5.4-12-12-12H96c-53 0-96 43-96 96v192c0 53 43 96 96 96h84c6.6 0 12-5.4 12-12z",
                ],
                "iconName": "sign-out-alt",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
           LOGOUT
        </Link>
      </li>
    </ul>
  </nav>,
]
`;
